src/libraries/Microsoft.PowerFx.Json/Functions/JsonFunctionImpl.cs

```diff
@@ -129,17 +143,58 @@ private JsonFlags GetFlags()
 
             private class Utf8JsonWriterVisitor : IValueVisitor
             {
+                private const int _maxDepth = 20;           // maximum depth of UO
+
+                private const int _maxLength = 1024 * 1024; // 1 MB, maximum number of bytes allowed to be sent to Utf8JsonWriter
+
                 private readonly Utf8JsonWriter _writer;
+
                 private readonly TimeZoneInfo _timeZoneInfo;
+
                 private readonly bool _flattenValueTables;
 
+                private readonly Canceller _canceller;
+
                 internal readonly List<ErrorValue> ErrorValues = new List<ErrorValue>();
 
-                internal Utf8JsonWriterVisitor(Utf8JsonWriter writer, TimeZoneInfo timeZoneInfo, bool flattenValueTables)
+                internal Utf8JsonWriterVisitor(Utf8JsonWriter writer, TimeZoneInfo timeZoneInfo, bool flattenValueTables, Canceller canceller)
                 {
                     _writer = writer;
                     _timeZoneInfo = timeZoneInfo;
                     _flattenValueTables = flattenValueTables;
+
+                    _canceller = canceller;
+                }
+
+                private void ThrowIfCancellationRequested(int index)
+                {
+                    _canceller.ThrowIfCancellationRequested();
+
+                    if (index > _maxDepth)
+                    {
+                        IRContext irContext = IRContext.NotInSource(FormulaType.UntypedObject);
+                        ErrorValues.Add(new ErrorValue(irContext, new ExpressionError()
+                        {
+                            ResourceKey = TexlStrings.ErrReachedMaxJsonDepth,
+                            Span = irContext.SourceContext,
+                            Kind = ErrorKind.InvalidArgument
+                        }));
+
+                        throw new InvalidOperationException("Maximum depth reached while traversing JSON payload.");
```

Comment:
>Maximum depth [](http://example.com/codeflow?start=61&length=13)

nit- it's good to include the actual max depth/length in the error message. 
This will be more interesting once those numbers are adjustable.

Reply:
good idea!